### PR TITLE
Support symbolic/hard link

### PR DESCRIPTION
**Pull request by kuozheyu**
Support symbolic/hard link

### CODE REVIEW

File: inode.c [hunk 1]
```diff
@@ -337,18 +351,27 @@ static int simplefs_unlink(struct inode *dir, struct dentry *dentry)
...
 
+    if (S_ISLNK(inode->i_mode))
+        goto clean_inode;
     /* Update inode stats */
     dir->i_mtime = dir->i_atime = dir->i_ctime = current_time(dir);
     if (S_ISDIR(inode->i_mode))
         inode_dec_link_count(dir);
     mark_inode_dirty(dir);
 
+
```
Comment (jserv) [inline inode.c]: Avoid this blank line.

File: inode.c [hunk 2]
```diff
@@ -533,11 +556,112 @@ static int simplefs_rmdir(struct inode *dir, struct dentry *dentry)
...
+            SIMPLEFS_FILENAME_LEN);
+    mark_buffer_dirty(bh);
+
+    inode_inc_link_count(inode);
+
+    d_instantiate(dentry, inode);
+end:
+    brelse(bh);
+    return ret;
+}
```
Comment (jserv) [inline inode.c:599]: Append extra blank line.

File: inode.c [hunk 3]
```diff
@@ -533,11 +556,112 @@ static int simplefs_rmdir(struct inode *dir, struct dentry *dentry)
...
+
+    inode->i_link = (char *) ci->i_data;
+    memcpy(inode->i_link, symname, l);
+    inode->i_size = l - 1;
+    mark_inode_dirty(inode);
+    d_instantiate(dentry, inode);
+
+    return 0;
+}
+const char *simplefs_get_link(struct dentry *dentry,
```
Comment (jserv) [inline inode.c]: This function should be declared as `static`.

File: inode.c [hunk 4]
```diff
@@ -533,11 +556,112 @@ static int simplefs_rmdir(struct inode *dir, struct dentry *dentry)
...
 static const struct inode_operations simplefs_inode_ops = {
     .lookup = simplefs_lookup,
     .create = simplefs_create,
     .unlink = simplefs_unlink,
     .mkdir = simplefs_mkdir,
     .rmdir = simplefs_rmdir,
     .rename = simplefs_rename,
-};
+    .link = simplefs_link,
+    .symlink = simplefs_symlink};
```
Comment (jserv) [inline inode.c]: Make sure that you already indent all source files with `clang-format -i`.

File: inode.c [hunk 5]
```diff
@@ -167,6 +171,16 @@ static struct inode *simplefs_new_inode(struct inode *dir, mode_t mode)
...
+
+    if (S_ISLNK(mode)) {
+        inode_init_owner(inode, dir, mode);
+        set_nlink(inode, 1);
+        inode->i_ctime = inode->i_atime = inode->i_mtime = current_time(inode);
+        inode->i_op = &simplefs_symlink_inode_ops;
+        return inode;
+    }
+
+
```
Comment (jserv) [inline inode.c]: Avoid adding this blank line.

File: inode.c [hunk 6]
```diff
@@ -192,7 +206,6 @@ static struct inode *simplefs_new_inode(struct inode *dir, mode_t mode)
     }
 
     inode->i_ctime = inode->i_atime = inode->i_mtime = current_time(inode);
-
```
Comment (jserv) [inline inode.c]: Don't do that. Minimize the functional changes as possible.

File: inode.c [hunk 7]
```diff
@@ -533,11 +556,112 @@ static int simplefs_rmdir(struct inode *dir, struct dentry *dentry)
...
+    mark_buffer_dirty(bh);
+
+    inode_inc_link_count(inode);
+
+    d_instantiate(dentry, inode);
+end:
+    brelse(bh);
+    return ret;
+}
+int simplefs_symlink(struct inode *dir,
```
Comment (jserv) [inline inode.c]: This function should be declared as `static`.

File: simplefs.h [hunk 8]
```diff
@@ -38,6 +38,7 @@ struct simplefs_inode {
     uint32_t i_blocks;    /* Block count */
     uint32_t i_nlink;     /* Hard links count */
     uint32_t index_block; /* Block with list of blocks for this file */
+    char i_data[32];      /*store symlink content*/
```
Comment (jserv) [inline simplefs.h]: The comment should be `/* store symlink content */`. Be aware of the spaces.

File: inode.c [hunk 9]
```diff
@@ -533,11 +556,112 @@ static int simplefs_rmdir(struct inode *dir, struct dentry *dentry)
...
+    struct simplefs_inode_info *ci = SIMPLEFS_INODE(inode);
+    struct simplefs_inode_info *ci_dir = SIMPLEFS_INODE(dir);
+    struct simplefs_dir_block *dir_block;
+    struct buffer_head *bh;
+    int f_pos = 0, i = 0;
+
+    if (l > sizeof(ci->i_data))
+        return -ENAMETOOLONG;
+
+    // fill directory data block
```
Comment (jserv) [inline inode.c]: Use C-style comments. That is `/* ... */` rather than `//` for consistency.

**Comment by kuozheyu** [comment]
Ok, I will resend again, and thank you for your review.

**Comment by jserv** [comment]
> Ok, I will resend again, and thank you for your review.

You don't have to send another pull request. Just modify file(s), use `git add; git commit --amend` and then `git push --force`.

**Comment by jserv** [comment]
By the way, rebasing the latest `master` branch is required.

**Comment by kuozheyu** [comment]
I got it, thanks.

File: inode.c [hunk 10]
```diff
@@ -144,9 +149,9 @@ static struct inode *simplefs_new_inode(struct inode *dir, mode_t mode)
     int ret;
 
     /* Check mode before doing anything to avoid undoing everything */
-    if (!S_ISDIR(mode) && !S_ISREG(mode)) {
+    if (!S_ISDIR(mode) && !S_ISREG(mode) && !S_ISLNK(mode)) {
         pr_err(
-            "File type not supported (only directory and regular files "
+            "File type not supported (only directory, regular and symlink files"
```
Comment (jserv) [inline inode.c]: Change "regular and symlink files" to "regular file and symlink"

File: inode.c [hunk 11]
```diff
@@ -533,11 +557,113 @@ static int simplefs_rmdir(struct inode *dir, struct dentry *dentry)
...
 static const struct inode_operations simplefs_inode_ops = {
     .lookup = simplefs_lookup,
     .create = simplefs_create,
     .unlink = simplefs_unlink,
     .mkdir = simplefs_mkdir,
     .rmdir = simplefs_rmdir,
     .rename = simplefs_rename,
-};
+    .link = simplefs_link,
+    .symlink = simplefs_symlink};
```
Comment (jserv) [inline inode.c]: Append extra `,` close to `.symlink = simplefs_symlink`, so that `clang-format` can indent as expected.

File: inode.c [hunk 12]
```diff
@@ -533,11 +557,113 @@ static int simplefs_rmdir(struct inode *dir, struct dentry *dentry)
...
+    struct super_block *sb = dir->i_sb;
+    unsigned int l = strlen(symname) + 1;
+    struct inode *inode = simplefs_new_inode(dir, S_IFLNK | S_IRWXUGO);
+    struct simplefs_inode_info *ci = SIMPLEFS_INODE(inode);
+    struct simplefs_inode_info *ci_dir = SIMPLEFS_INODE(dir);
+    struct simplefs_dir_block *dir_block;
+    struct buffer_head *bh;
+    int f_pos = 0, i = 0;
+
+    if (l > sizeof(ci->i_data))
```
Comment (jserv) [inline inode.c:615]: Shall we mention the limitation of symlink?

**Comment by kuozheyu** [comment]
Ok

File: inode.c [hunk 13]
```diff
@@ -533,11 +557,115 @@ static int simplefs_rmdir(struct inode *dir, struct dentry *dentry)
...
     .unlink = simplefs_unlink,
     .mkdir = simplefs_mkdir,
     .rmdir = simplefs_rmdir,
     .rename = simplefs_rename,
+    .link = simplefs_link,
+    .symlink = simplefs_symlink,
 };
+
+static const struct inode_operations simplefs_symlink_inode_ops = {
+    .get_link = simplefs_get_link};
```
Comment (jserv) [inline inode.c]: Append `,` right after `simplefs_get_link` to make consistent indention.

File: inode.c [hunk 14]
```diff
@@ -9,6 +9,7 @@
 #include "simplefs.h"
 
 static const struct inode_operations simplefs_inode_ops;
+static const struct inode_operations simplefs_symlink_inode_ops;
```
Comment (jserv) [inline inode.c]: Shorten it to `symlink_inode_ops` for simplicity.

**Comment by jserv** [comment]
Thank @kuozheyu for contributing!

**Comment by kuozheyu** [comment]
Thanks for your help.

**Pull request merged**